staging/src/k8s.io/client-go/tools/watch/informerwatcher.go

```diff
@@ -127,7 +127,7 @@ func NewIndexerInformerWatcher(lw cache.ListerWatcher, objType runtime.Object) (
 				// We have no means of passing the additional information down using
 				// watch API based on watch.Event but the caller can filter such
 				// objects by checking if metadata.deletionTimestamp is set
-				obj = staleObj
```

Comment:
Can you update a test? This problem shouldn't be possible if a test exercised this even once.

Reply:
I don't think that is possible without a refactor because [all the fake types that facilitate unit-testing this function](https://github.com/kubernetes/kubernetes/blob/46d481b4556e3328758eeb851920e2536a488658/staging/src/k8s.io/client-go/tools/watch/informerwatcher_test.go#L219-L221) only deal with `runtime.Object` and I can't find how to simulate a `DeltaFIFO` sending `cache.DeletedFinalStateUnknown` into the `DeleteFunc` without changing how `NewIndexerInformerWatcher` works.

If refactoring is ok, we could declare `var newIndexInformer = cache.NewIndexerInformer` globally in the package scope, then replace it with a mock in the unit test with a function that will allow us to send whatever we want. I'm not familiar with the code base so I might be missing something here.

Any pointer would be greatly appreciated.

Reply:
You can drive it with a specially crafted listerwatcher that returns a list containing an item on the first List() invocation, and a list with no items on subsequent List() invocations, and a Watch() implementation that returns no events:

Add to staging/src/k8s.io/client-go/tools/watch/informerwatcher_test.go:
```go

func TestInformerWatcherDeletedFinalStateUnknown(t *testing.T) {
	listCalls := 0
	watchCalls := 0
	lw := &cache.ListWatch{
		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
			retval := &corev1.SecretList{}
			if listCalls == 0 {
				// return a list with items in it
				retval.ResourceVersion = "1"
				retval.Items = []corev1.Secret{{ObjectMeta: metav1.ObjectMeta{Name: "secret1", Namespace: "ns1", ResourceVersion: "123"}}}
			} else {
				// return empty lists after the first call
				retval.ResourceVersion = "2"
			}
			listCalls++
			return retval, nil
		},
		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
			w := watch.NewFake()
			if options.ResourceVersion == "1" {
				go func() {
					// close with a "Gone" error when trying to start a watch from the first list
					w.Error(&apierrors.NewGone("gone").ErrStatus)
					w.Stop()
				}()
			}
			watchCalls++
			return w, nil
		},
	}
	_, _, w, done := NewIndexerInformerWatcher(lw, &corev1.Secret{})

	// Expect secret add
	select {
	case event, ok := <-w.ResultChan():
		if !ok {
			t.Fatal("unexpected close")
		}
		if event.Type != watch.Added {
			t.Fatalf("expected Added event, got %#v", event)
		}
		if event.Object.(*corev1.Secret).ResourceVersion != "123" {
			t.Fatalf("expected added Secret with rv=123, got %#v", event.Object)
		}
	case <-time.After(time.Second * 10):
		t.Fatal("timeout")
	}

	// Expect secret delete because the relist was missing the secret
	select {
	case event, ok := <-w.ResultChan():
		if !ok {
			t.Fatal("unexpected close")
		}
		if event.Type != watch.Deleted {
			t.Fatalf("expected Deleted event, got %#v", event)
		}
		if event.Object.(*corev1.Secret).ResourceVersion != "123" {
			t.Fatalf("expected deleted Secret with rv=123, got %#v", event.Object)
		}
	case <-time.After(time.Second * 10):
		t.Fatal("timeout")
	}

	if listCalls < 2 {
		t.Fatalf("expected at least 2 list calls, got %d", listCalls)
	}
	if watchCalls < 1 {
		t.Fatalf("expected at least 1 watch call, got %d", watchCalls)
	}

	w.Stop()
	select {
	case <-done:
	case <-time.After(time.Second * 10):
		t.Fatal("timeout")
	}
}
```

Reply:
Ah, I see. So the key here is trying to simulate a scenario described in the comment here:
https://github.com/kubernetes/kubernetes/blob/dc39ab2417bfddcec37be4011131c59921fdbe98/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L736-L739

Thanks, @liggitt! You're awesome!